src/python/pants/backend/python/rules/resolve.py

```diff
@@ -0,0 +1,356 @@
+async def create_resolver(pex_bin: PexBin) -> Resolver:
+    resolver_snapshot = await Get[Snapshot](InputFilesContent, Resolver.files())
+    init_injected = await Get[InitInjectedSnapshot](InjectInitRequest(snapshot=resolver_snapshot))
+    return Resolver(
+        pex=await Get[Pex](
+            PexRequest(
+                output_filename="bin/resolver.pex",
+                requirements=PexRequirements((pex_bin.requirement, "dataclasses==0.6")),
+                sources=init_injected.snapshot.digest,
+                entry_point=Resolver.entry_point(),
+                additional_args=("--no-strip-pex-env", "--unzip"),
+            )
+        )
+    )
+
+
+@dataclass(frozen=True)
+class PexCache:
+    snapshot: Snapshot = EMPTY_SNAPSHOT
+    path: str = "pex_root"
+
+
+@dataclass(frozen=True)
+class ResolveRequest:
+    requirements: PexRequirements
+    cache: PexCache = PexCache()
+    local: bool = False
+    requirer: Optional[str] = None
+
+
+@dataclass(frozen=True)
+class ResolveResult:
+    dependencies: DistributionDependencies = DistributionDependencies()
+    cache: PexCache = PexCache()
+
+
+@named_rule(desc="Resolve python requirements")
+async def resolve_requirements(
+    resolver: Resolver,
+    python_setup: PythonSetup,
+    python_repos: PythonRepos,
+    request: ResolveRequest,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment,
+    log_level: LogLevel,
+) -> ResolveResult:
+    argv = list(PexDebug(log_level).iter_pex_args())
+
+    if python_setup.requirement_constraints is not None:
+        constraint_file_snapshot = await Get[Snapshot](
+            PathGlobs(
+                [python_setup.requirement_constraints],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                conjunction=GlobExpansionConjunction.all_match,
+                description_of_origin="the option `--python-setup-requirement-constraints`",
+            )
+        )
+        argv.extend(["--constraints", python_setup.requirement_constraints])
+    else:
+        constraint_file_snapshot = EMPTY_SNAPSHOT
+
+    for interpreter_search_path in python_setup.interpreter_search_paths:
+        argv.extend(["--interpreter-search-path", interpreter_search_path])
+
+    for interpreter_constraint in python_setup.interpreter_constraints:
+        argv.extend(["--interpreter-constraint", interpreter_constraint])
+
+    if python_setup.resolver_jobs:
+        argv.extend(["--jobs", python_setup.resolver_jobs])
+
+    if python_setup.manylinux:
+        argv.extend(["--manylinux", python_setup.manylinux])
+    else:
+        argv.append("--no-manylinux")
+
+    if python_setup.resolver_allow_prereleases:
+        argv.append("--pre")
+
+    for platform in python_setup.platforms:
+        argv.extend(["--platform", platform])
+
+    if python_repos.indexes:
+        for index in python_repos.indexes:
+            argv.extend(["--index", index])
+    else:
+        argv.extend("--no-index")
```

Comment:
In `pex.py`, we always set `--no-index` (`--no-pypi`) so that users can exclude it, when desired. Why not use the same snippet?

https://github.com/pantsbuild/pants/blob/766ce1d0127a3c59afb6d8cca533616fa427e57d/src/python/pants/backend/python/rules/pex.py#L279-L283